transactions/src/main/java/org/jboss/as/txn/suspend/RecoverySuspendController.java

```diff
@@ -48,9 +64,45 @@ public void suspended(ServerActivityCallback serverActivityCallback) {
         serverActivityCallback.done();
     }
 
+    /**
+     * {@link RecoveryManagerService#resume() Resumes} the {@link RecoveryManagerService} if the current
+     * process state {@link ControlledProcessState.State#isRunning() is running}. Otherwise records that
+     * the service can be resumed once a {@link #propertyChange(PropertyChangeEvent) notification is received} that
+     * the process state is running.
+     */
     @Override
     public void resume() {
-        recoveryManagerService.resume();
+        boolean doResume;
+        synchronized (this) {
+            suspended = false;
+            doResume = running;
+        }
+        if (doResume) {
+            resumeRecovery();
+        }
     }
 
+    /**
+     * Receives notifications from a {@code ProcessStateNotifier} to detect when the process has reached a
+     * {@link ControlledProcessState.State#isRunning()}  running state}, reacting to them by
+     * {@link RecoveryManagerService#resume() resuming} the {@link RecoveryManagerService} if we haven't been
+     * {@link #preSuspend(ServerActivityCallback) suspended}.
+     */
+    @Override
+    public void propertyChange(PropertyChangeEvent evt) {
+        boolean doResume;
+        synchronized (this) {
+            ControlledProcessState.State newState = (ControlledProcessState.State) evt.getNewValue();
+            running = newState.isRunning();
+            doResume = running && !suspended;
+        }
+        if (doResume) {
+            resumeRecovery();
+        }
+
+    }
```

Comment:
@mmusgrov @ochaloup  FYI. What this is about is the fact that with this RFE the behavior of boot can be different, where users can configure 'graceful startup' to be turned off. With graceful startup this kind of sequence happens

server is initially in a suspended state as boot proceeds
boot is 99% done, boot logic 'resumes' the server to move out of suspended. The resume() method above is called
server switches from ControlledProcessState.State.STARTING to RUNNING (i.e. boot is done)

The gap between the last two bits is miniscule.

If the user has disable graceful startup, the server is not initially in suspended state, so no resume call goes out. So no signal that triggers 'recoveryManagerService.resume()' will be received. So this change is adding a listener for when the ControlledProcessState.State.STARTING to RUNNING change happens and uses that to ensure recoveryManagerService.resume() is called.

If after boot the server is suspend and later resumed, that will result in a call to recoveryManagerService.resume(), same as before.